### ISSUE TITLE
Regex Escaping

### ISSUE THREAD

**Issue by absherwin** [JuliaLang/julia#6124]
I can't find a function to escape a string for use in regular expression matching. That is: f such that match(Regex(f(s1)),s2)!=nothing iff s1==s2. 

Python has such a function (re.escape) as do Ruby (Regexp.escape) and MATLAB (regextranslate).  R, on the the other hand does not. 

Assuming the community decides to add, escape_regex seems like a clear name but bothers me slightly because it's acting on a string rather than a Regex. escape_regexstring seems more accurate but needlessly verbose. 

I'm happy to contribute the patch if the community agrees this is worthwhile. 

Thoughts?

**Comment by bkamins** [comment]
@StefanKarpinski The following code
```
s1 = "\\E"
r = Regex("^\\Q$s1\\E\$")
match(r, "")
```
finds a match, however it should not as we wanted to look for *\E*. Maybe it is worth to consider reopening this issue and adding a function that escapes string for use in regex?

**Comment by StefanKarpinski** [comment]
Yeah, this could use a more complete solution. It seems to be strictly a feature addition, however.

**Comment by mlhetland** [comment]
Didn't want to go full PR on this (yet), but it seems there are semi-standard ways of doing this. My implementations of three canonical-ish versions follow. First, the [Perl version](http://perldoc.perl.org/functions/quotemeta.html), which escapes everything except word characters:

```julia
# Perl version:
quotemeta(s::AbstractString) = replace(s, r"(\W)" => s"\\\1")
```

As the docs say, this is the internal version behind the `\Q` escapes; given that those are available in PCRE as well, I assume there might be a function there we could expose directly (though a cursory grepping through the code didn't uncover it).

This is simple enought it might simply be a recipe in the docs, I guess. `¯\_(ツ)_/¯`

The C++ wrapper of PCRE has [an implementation of this](https://github.com/vmg/pcre/blob/f2b1a56ddb0a3b9ffe42f45fde2a0d1318eadf90/pcrecpp.cc) (which suggests that it's *not* available directly from PCRE), and it does the same thing, except it also replaces `"\0"` with `"0"` (in addition to escaping it), as PCRE doesn't handle `"\0"`, escaped or not:

```julia
# pcrecpp version:
function quotemeta(s::AbstractString)
    res = replace(s, r"(\W)" => s"\\\1")
    replace(res, "\0" => "0")
end
```

Finally, there's the [more conservative PHP version](http://php.net/manual/en/function.preg-quote.php), which only escapes the special characters `.\+*?[^]$(){}=!<>|:-`, and possibly one optional delimiter (which I guess ought to be a single character, but the PHP docs don't explicitly state that; then again, escaping the first character of a character sequence might be enough).

```julia
# PHP version:
function quotemeta(s::AbstractString; delim=nothing)
    res = replace(s, r"([.\\+*?[^\]$(){}=!<>|:-])" => s"\\\1")
    delim ≢ nothing ? replace(res, delim => "\\$delim") : res
end
```

I do think escaping `"\0"` seems sensible, as in pcrecpp – and I also think escaping «everything» is perhaps going a bit far (though it does ensure that no special character is left out, of course). On the other hand, the extra delimiter seems a bit out of place (for us). Maybe a combination of the pcrecpp and PHP versions?

```julia
# Julia version?
function quotemeta(s::AbstractString)
    res = replace(s, r"([.\\+*?[^\]$(){}=!<>|:-])" => s"\\\1")
    replace(res, "\0" => "\\0")
end
```

**Comment by mlhetland** [comment]
[The Python version](https://github.com/python/cpython/blob/3.7/Lib/re.py) escapes `()[]{}?*+-|^$\\.&~# \t\n\r\v\f`; from what I can tell, [the Ruby version](https://sonots.github.io/ruby-capi/dc/d26/re_8c_source.html#l03425) omits`&` and `~`. [The MATLAB version](https://www.mathworks.com/help/matlab/ref/regexptranslate.html) seems to be customizable, but also follows the strategy of selectively escaping special characters (rather than simply `\W`).

**Comment by mlhetland** [comment]
Escaping whitespace does seem to make sense (though we might as well use `\s`) – as it'll be ignored if we're using the `x` modifier. Same with `#`, for comments.

**Comment by mlhetland** [comment]
So, including these extra precautionary characters from the Python version (but escaping *all* whitespace, not just the characters listed):

```julia
function quotemeta(s::AbstractString)
    res = replace(s, r"([()[\]{}?*+\-|^\$\\.&~#\s=!<>|:])" => s"\\\1")
    replace(res, "\0" => "\\0")
end
```

**Comment by mlhetland** [comment]
… though I guess if PCRE *doesn't* ignore other whitespace than the above in `x`-mode, then the use of `\s` is just an unnecessary performance hit, and we might as well list the specific offenders.

**Comment by amellnik** [comment]
@mlhetland Did you want to make a PR for this?  If not, I can.  Cheers!

**Comment by mlhetland** [comment]
Feel free to!-) I’d suggest looking into that last issue – i.e. what PCRE ignores in x-mode vs what is matched by `\s`. If the latter is a strict superset, we might just go with the former (listed explicitly).

**Comment by erezsh** [comment]
Ping! Why is the PR still on hold?

**Comment by StefanKarpinski** [comment]
See also: https://github.com/JuliaLang/julia/pull/29643

**Comment by erezsh** [comment]
I don't understand the practice of closing an issue before the PR has been merged.

**Comment by mlhetland** [comment]
I think the point is that there is another function (in Base) that solves the problem, even though it's up in the air whether the one from the PR should be added in addition.

**Comment by mlhetland** [comment]
Se, specifically, [this comment](https://github.com/JuliaLang/julia/pull/29643#pullrequestreview-580658197), about `Base.wrap_string`, which provides the functionality that is the topic of this issue.

**Comment by erezsh** [comment]
That comment was left two days ago, for a PR that's been open since 2018. `wrap_string` isn't even documented yet.

Does `wrap_string` escape things like `[` or `.`? If not, then it's nothing like regex escaping.

It's another thing if you said "we don't care about this feature." Then it would make more sense to me.

**Comment by mlhetland** [comment]
I have nothing to do with the decisions, here. I just implemented the function in the PR, because I also needed it (two years ago). Just tried to explain my impression of why the issue was closed – but I agree that until `wrap_string` is at least documented, the issue could well be left open.

As for what `wrap_string` does, I *think* it uses the Perl approach of regexp escape, which essentially escapes everything; but it does seem mostly for internal use (with a mysterious second integer argument…).
